models/kernels/Nonlinear_Reservoir.hpp

```diff
@@ -0,0 +1,153 @@
+        state.current_storage_height_meters = current_storage_height_meters;
+    }
+
+    //Constructor for a reservoir with only one outlet.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters, double a, 
+    double b, double activation_threshold_meters) : Nonlinear_Reservoir(minimum_storage_meters, maximum_storage_meters, 	current_storage_height_meters)
+    {
+        //Ensure that the activation threshold is less than the maximum storage
+        //if (activation_threshold_meters > maximum_storage_meters)
+        //TODO: Return appropriate error
+
+        this->outlets.push_back(Reservoir_Outlet(a, b, activation_threshold_meters));
+    }
+
+    //Constructor for a reservoir with multiple outlets.
+    Nonlinear_Reservoir(double minimum_storage_meters, double maximum_storage_meters, double current_storage_height_meters, std::vector <Reservoir_Outlet> &outlets) : Nonlinear_Reservoir(minimum_storage_meters, maximum_storage_meters, current_storage_height_meters)
+    {
+
+        this->outlets = outlets;
+	  
+        //Ensure that reservoir outlet activation thresholds are sorted from least to greatest height
+        sort(this->outlets.begin(), this->outlets.end(), [](const Reservoir_Outlet &left, const Reservoir_Outlet &right)
+        {              
+            return left.activation_threshold_meters < right.activation_threshold_meters;
+        });
+
+        //Ensure that the activation threshold is less than the maximum storage
+        for (auto& outlet : this->outlets) //pointer to outlets
+        {
+            if (outlet.activation_threshold_meters > maximum_storage_meters)
+            //TODO: Return appropriate error
+                continue; 
+        }
+
+    }	
+
+    //Function to update the response of storage in meters to an influx and timestep.
+    double response_storage_meters(double in_flux_meters_per_second, int delta_time_seconds)
+    {	
+        state.current_storage_height_meters += in_flux_meters_per_second * delta_time_seconds;
+	
+        for (auto& outlet : this->outlets) //pointer to outlets
+        {
+            //Update storage from outlet velocity multiplied by delta time.
+            state.current_storage_height_meters -= outlet.velocity_meters_per_second(parameters, state) * delta_time_seconds;
+        }
+
+        //If storage is less than minimum storage, correct to minimum storage.
+        if (state.current_storage_height_meters < parameters.minimum_storage_meters)
+            state.current_storage_height_meters = parameters.minimum_storage_meters;
+
+        //If storage is greater than maximum storage, correct to maximum storage.
+        if (state.current_storage_height_meters > parameters.maximum_storage_meters)
+            state.current_storage_height_meters = parameters.maximum_storage_meters;   
+            //TODO: Return extra storage to caller	
```

Comment:
this would be (current - max)/dt

I think passing in_flux by reference and taking from the flux and returning to it is the most straight forward way to handle this.  The caller must be prepared to check pass a flux, and then check that it has been exhausted (i.e. is 0) or has some left (i.e. > 0) and do something with it.

@donaldwj what do you think about this method?

Reply:
Oh right thanks. I will change to (current - max)/dt. @jdmattern-noaa fixing.